#381: update only what was changed

Garagoth opened the pull request:
Please review and comment. I might have missed some cases - please see Defect381_UpdateModifiedOnlyTest.java for covered scenarios.

ericbn commented:
@Garagoth, I just modified the Convert code to be locale independent, following the issue #384. With this new code the test should be:

    //try numeric string with decimal separator in java format, this should be locale independent
    a.set("amount", "123.11");
    a.validate();
    a(a.errors().size()).shouldBeEqual(0);

I'm suggesting we use [Converters](http://javalite.io/data_conversions#converters) to do locale-specific conversions between numbers and strings.

ipolevoy commented:
@ericbn, are you saying that this pull request is not necessary?

ericbn commented:
@ipolevoy, @Garagoth, the changes made to NumericValidator are not necessary.

ericbn commented:
@Garagoth, also please add a test that does `delete()`, then `thaw()`, then `saveIt()` on the model. It should be "modified" between `thaw()` and `saveIt()`, right?

Garagoth commented:
@ericbn : I added to numericValidator comment regarding Locales AND some speedup with skipping .toString() + .parse() combo when value is instanceof Number already.
Main change was in test to be include locale specific string...

Also, regarding isModified: Yes, it should be modified between thaw and freezeit, give me a moment :-)

Garagoth commented:
@ericbn : Added tests.
Also, regarding NumericValidator: after your change it still fails with decimal separator different then locale settings.

ericbn commented:
@Garagoth, `org.javalite.common.Convert` only accepts the java format for numbers now, so only `'.'` for decimal separator. That's what I meant for "locale independent" above at https://github.com/javalite/activejdbc/pull/386#issuecomment-79091611, because now the behavior won't vary automatically with the server locale.

I'm suggesting we use `org.javalite.activejdbc.conversion.Converter` to do locale-specific conversions between numbers and strings.

So we need to review NumericValidator with the new behavior -- in case of an String, accept only the java format for numbers now, so only `'.'` for decimal separator, disregarding the server locale.

As this issue is not the target of this pull request, I think we can review this latter...

Garagoth commented:
@ericbn As I said, I made no functional changes in NumericValidator other than some speed improvement for objects that are already numbers.
I modified the test to show the issue and to make it pass on my machine as it wasn't working other way. It should pass everywhere now, regardless of locale used.
Can we deal with updates now and move to discussing and fixing validators in separate thread?

ericbn commented:
@Garagoth, sure! @ipolevoy, this pull request LGTM.

Garagoth commented:
@ipolevoy I changed tests to conform to naming conventions. Also added some tests for dirty list contents.
Regarding passing dirty info to new model in copyTo: since you put directly into attributes those became changed, so dirty. Imagine the scenario:
m1.createIt(); m2.createIt(); m1.copyTo(m2);
m2 is not new but modified. I created a test for that: shouldBeModifiedAfterCopyTo()
